### DISCUSSION TITLE
SimpleTcpServer DataReceived method spawns a new task for each DataReceived event

### DISCUSSION THREAD

**Discussion by TheNybbler** [jchristn/SuperSimpleTcp#161] [General]
The implementation of the `DataReceiver` method in the `SimpleTcpServer` spawns a new Task to call `_events.HandeDataReceived`. this could potentially lead to problems as there is no guarantee that the tasks will run in the correct order (in high data rate scenarios). As each client runs the `DataReceiver` as a separate task, I don't think it is necessary to do this (unless I'm missing something?)

Interestingly, the `SimpleTcpClient` equivalent method does NOT spawn a new Task to call the `_events.HandleDataReceived` method, which is inconsistent.

**Comment by jchristn** [comment]
The client doesn't need to. It's managing exactly one TCP connection. The server manages many which is why a task is spawned for each. I did not want to implement my own scheduler so I left that to be the job of the underlying operating system.

**Comment by TheNybbler** [comment]
But each server client connection already spawns a task to run its own data receiver. I don't see the benefit of then also spawning a task to handle the firing of the data received event.

**Comment by jchristn** [comment]
Those events are fired synchronously. By burying them in a task it allows the server to fire more in parallel and get back to reading from the socket immediately.

**Comment by TheNybbler** [comment]
It would nice to have the option to control this behaviour. In my use case I only have a single client connecting to the server at a time and high speed continous throughput is important. By having the read events firing from within tasks, this introduces the possibility of getting out of sequence data when reading as there is no guarantee as to which task will be scheduled next.

**Reply by jchristn** [reply]
Sure thing.  Happy to take a PR on this if you'd like to put one together.  Cheers

**Reply by TheNybbler** [reply]
I've created a PR with the changes along with a couple of unit tests.

**Reply by jchristn** [reply]
Thanks @TheNybbler - will look at it ASAP!  Much appreciated!
